lib/component.js

```diff
@@ -805,6 +819,114 @@ class Component extends WebComponent {
     return this._attrs;
   }
 
+  _syncParams() {
+    // maintain local validated map where all schema keys are defined
+    this._paramTypes = {};
+    const paramTypes = this.getConfig(`params`);
+
+    // vuejs like this.data.$data to reference "real" value of params
+    Object.defineProperty(this.params, `$params`, {
+      enumerable: false,
+      value: {},
+    });
+
+    for (const param of Object.keys(paramTypes)) {
+      // convert type shorthand to object
+      let paramType = paramTypes[param];
+      if (PARAM_TYPE_DEFAULTS.has(paramType)) {
+        paramType = {type: paramType};
+      }
+
+      // Ensure param type is valid
+      const type = paramType.type;
+      if (!PARAM_TYPE_DEFAULTS.has(type)) {
+        const typeString = typeof type === `function` ? type.name : String(type);
+        throw new Error(
+          `Invalid type: ${typeString} for attr: ${param} in paramType. ` +
+            `Only (${Array.from(PARAM_TYPE_DEFAULTS.keys())
+              .map((v) => `'${v.name}'`)
+              .join(` | `)}) is valid.`,
+        );
+      }
+
+      if (paramType.default && paramType.required) {
+        throw new Error(`${this}: param '${param}' cannot have both required and default`);
+      }
+
+      if ([Object, Array, Map, Set].includes(type) && !paramType.shouldUpdate) {
+        throw new Error(
+          `Object like params '${param}' on ${this.constructor.name} must have 'shouldUpdate' function passed in`,
+        );
+      }
+
+      const paramSchemaObj = {
+        type,
+        default: paramType.default ? paramType.default : PARAM_TYPE_DEFAULTS.get(type),
```

Comment:
`ParamSchema.default` is allowed to be undefined, it should not fallback to `PARAM_TYPE_DEFAULTS.get(type)`. When the default is undefined, it means that the param is required and will cause errors if the component’s params are updated without a setting a value for this param.

Reply:
just add explicit check for `paramSchema.default !== undefined` for this. it will fallback to empty value for those primitive types only if no `default` key is provided to the param schema. 

technically when a value is declared then it's value won't be undefined, at least it should be `null`.

Reply:
Hm thinking about it a bit more, I think this part of the Panel Params spec is wacky. There shouldn't be a `default` for each individual param schema, defaults should be done the same way React does it and how Panel does default states: a single dictionary containing all defaults.

Default values should be declared all at once in config like this:
```typescript
get config() {
  return {
    ...
    defaultParams: {
      x: 1,
      y: 2,
    },
  };
}
```

Then each param schema should only have a `required` boolean, no `default`.

Reply:
hmm could do it to keep close to react.

Reply:
But either way, there should be no defaults for default values. If they don't specify a default for a certain param, the value _should be undefined_. It's not clear that there are always good defaults for parameters, for eg a component may work differently if a function param is not specified, just quietly filling in a no-op function as a supposedly "sensible default" is not desirable behavior.